src/core/runtime/shard.cc

```diff
@@ -121,7 +121,18 @@ class LegateShardingFunctor : public Legion::ShardingFunctor {
     auto hi    = proj_functor_->project_point(launch_space.hi(), launch_space);
     auto point = proj_functor_->project_point(p, launch_space);
 
-    auto shard_id = (linearize(lo, hi, point) + offset_) / per_node_count_ + start_node_id_;
+    auto tasks_per_node     = per_node_count_;
+    auto diff               = launch_space.get_volume() - per_node_count_ * total_shards;
+    uint32_t tasks_per_proc = 1;
+    if (diff > 0) {
+      auto num_shards      = end_node_id_ - start_node_id_;
+      uint32_t tasks_count = (launch_space.get_volume() + num_shards - 1) / num_shards;
+      tasks_per_node       = std::max(per_node_count_, tasks_count);
+      tasks_per_proc       = (tasks_per_node + per_node_count_ - 1) / per_node_count_;
+    }
+    auto proc_id  = (linearize(lo, hi, point) + offset_) / tasks_per_proc + start_node_id_;
+    auto shard_id = proc_id / per_node_count_;
+
```

Comment:
if you changed `LegateShardingFunctor` to be instantiated with a processor range (say `start_proc_id_` and `end_proc_id_`), instead of a pair of start and end nod ids, you could have used the same logic as `slice_task ` here; i.e.,
```
  uint32_t task_count = linearize(lo, hi, hi) + 1;
  uint32_t proc_count = end_proc_id_ - start_proc_id_ + 1;
  uint32_t idx = (linearize(lo, hi, p) * proc_count) / task_count + start_proc_id_;
  auto shard_id = global_proc_id / per_node_count_;
```